gusto/transport_equation.py

```diff
@@ -0,0 +1,345 @@
+              None, "once" or "twice". Defaults to "once".
+    :arg continuity: is the equation in continuity form L(q) = div(u*q)?
+                     If False, the equation is assumed to be in advective
+                     form L(q) = u dot grad(q)
+                     Defaults to False
+    """
+    def __init__(self, state, V, ibp="once", continuity=False):
+        super(Advection, self).__init__(state, V, ibp)
+        self.continuity = continuity
+
+    def advection_term(self, q):
+
+        if self.continuity:
+            if self.ibp is "once":
+                L = -inner(grad(self.test), outer(q, self.ubar))*dx
+            else:
+                L = inner(self.test, div(outer(q, self.ubar)))*dx
+        else:
+            if self.ibp is "once":
+                L = -inner(div(outer(self.test,self.ubar)),q)*dx
+            else:
+                L = inner(outer(self.test,self.ubar),grad(q))*dx
+
+        if self.dS is not None and self.ibp is not None:
+            L += dot(jump(self.test), (self.un('+')*q('+')
+                                       - self.un('-')*q('-')))*self.dS
+            if self.ibp is "twice":
+                L -= (inner(self.test('+'),dot(self.ubar('+'), self.n('+'))*q('+'))
+                      + inner(self.test('-'),dot(self.ubar('-'),
+                                                 self.n('-'))*q('-')))*self.dS
+
+        return L
+
+
+class EmbeddedDGAdvection(Advection):
+    """
+    Class for the transport equation, using an embedded DG advection scheme.
+
+    :arg state: :class:`.State` object.
+    :arg V: :class:`.FunctionSpace object. The function space that q lives in.
+    :arg ibp: (optional) string, stands for 'integrate by parts' and can take
+              the value None, "once" or "twice". Defaults to "once".
+    :arg continuity: (optional) Is the equation in continuity form
+                     L(q) = div(u*q)?
+                     If False, the equation is assumed to be in advective
+                     form L(q) = u dot grad(q)
+                     Defaults to False
+    :arg Vdg: (optional) :class:`.FunctionSpace object. The embedding function
+              space. Defaults to None which means that a broken space is
+              constructed for you.
+    """
+
+    def __init__(self, state, V, ibp="once", continuity=False, Vdg=None):
+
+        if Vdg is None:
+            # Create broken space, functions and projector
+            V_elt = BrokenElement(V.ufl_element())
+            space = FunctionSpace(state.mesh, V_elt)
+        else:
+            space = Vdg
+        self.xdg_out = Function(space)
+        self.x_projected = Function(V)
+        parameters = {'ksp_type':'cg',
+                      'pc_type':'bjacobi',
+                      'sub_pc_type':'ilu'}
+        self.Projector = Projector(self.xdg_out, self.x_projected,
+                                   solver_parameters=parameters)
+
+        super(EmbeddedDGAdvection, self).__init__(state, space, ibp)
```

Comment:
Miklos would say "call the superclass init first, so we're sure that there's no cross-contamination (in terms of needing to have set stuff up").  i.e. move this `super` call to after the space creation.